Assets/Mirror/Tests/WeaverTest.cs

```diff
@@ -0,0 +1,1012 @@
+//#define LOG_WEAVER_OUTPUTS
+
+using System.Collections.Generic;
+using NUnit.Framework;
+using UnityEngine;
+using UnityEditor.Compilation;
+
+using Mirror.Weaver;
+
+namespace Mirror
+{
+    [TestFixture]
+    public class WeaverTest
+    {
+        #region Private
+        List<string> m_weaverErrors = new List<string>();
+        void HandleWeaverError(string msg)
+        {
+#if LOG_WEAVER_OUTPUTS
+            Debug.LogError(msg);
+#endif
+            m_weaverErrors.Add(msg);
+        }
+
+        List<string> m_weaverWarnings = new List<string>();
+        void HandleWeaverWarning(string msg)
+        {
+#if LOG_WEAVER_OUTPUTS
+            Debug.LogWarning(msg);
+#endif
+            m_weaverWarnings.Add(msg);
+        }
+
+        private void BuildAndWeaveTestAssembly(string baseName)
+        {
+            WeaverAssembler.OutputFile = baseName + ".dll";
+            WeaverAssembler.AddSourceFiles(new string[] { baseName + ".cs" });
+            WeaverAssembler.Build();
+
+            Assert.That(WeaverAssembler.CompilerErrors, Is.False);
+            if (m_weaverErrors.Count > 0)
+            {
+                Assert.That(m_weaverErrors[0], Does.StartWith("Mirror.Weaver error: "));
+            }
+        }
+        #endregion
+
+        #region Setup and Teardown
+        [OneTimeSetUp]
+        public void FixtureSetup()
+        {
+            // TextRenderingModule is only referenced to use TextMesh type to throw errors about types from another module
+            WeaverAssembler.AddReferencesByAssemblyName(new string[] { "UnityEngine.dll", "UnityEngine.CoreModule.dll", "UnityEngine.TextRenderingModule.dll", "Mirror.dll" });
+
+            CompilationFinishedHook.UnityLogEnabled = false;
+            CompilationFinishedHook.OnWeaverError += HandleWeaverError;
+            CompilationFinishedHook.OnWeaverWarning += HandleWeaverWarning;
+        }
+
+        [OneTimeTearDown]
+        public void FixtureCleanup()
+        {
+            CompilationFinishedHook.UnityLogEnabled = true;
+        }
+
+        [SetUp]
+        public void TestSetup()
+        {
+        }
+
+        [TearDown]
+        public void TestCleanup()
+        {
+            WeaverAssembler.DeleteOutputOnClear = true;
+            WeaverAssembler.Clear();
+
+            m_weaverWarnings.Clear();
+            m_weaverErrors.Clear();
+        }
+        #endregion
+
+        #region General tests
+        [Test] // -----------------------------------------------------------------------------------------
+        public void InvalidType()
+        {
+            BuildAndWeaveTestAssembly(TestContext.CurrentContext.Test.Name);
```

Comment:
Is it possible to move this line to the [SetUp] method?

Reply:
https://github.com/vis2k/Mirror/pull/490/commits/36170bfe36fe8a86d35843bb6a23d858f6194265